exchanges/poloniex/poloniex_test.go

```diff
@@ -142,900 +84,2833 @@ func TestGetFee(t *testing.T) {
+
+	result, err = e.SubmitOrder(generateContext(t), &order.Submit{
+		Exchange:   e.Name,
+		Pair:       futuresTradablePair,
+		Side:       order.Buy,
+		MarginType: margin.Multi,
+		Type:       order.Limit,
+		Price:      10,
+		Amount:     10000000,
+		AssetType:  asset.Futures,
+	})
+	assert.NoError(t, err)
+	assert.NotNil(t, result)
+}
+
+func TestWebsocketSubmitOrder(t *testing.T) {
+	t.Parallel()
+
+	_, err := e.WebsocketSubmitOrder(t.Context(), nil)
+	require.ErrorIs(t, err, order.ErrSubmissionIsNil)
+
+	_, err = e.WebsocketSubmitOrder(t.Context(), &order.Submit{})
+	require.ErrorIs(t, err, common.ErrExchangeNameNotSet)
+
+	arg := &order.Submit{Exchange: e.Name}
+	_, err = e.WebsocketSubmitOrder(t.Context(), arg)
+	require.ErrorIs(t, err, order.ErrPairIsEmpty)
+
+	arg.Pair = spotTradablePair
+	_, err = e.WebsocketSubmitOrder(t.Context(), arg)
+	require.ErrorIs(t, err, order.ErrAssetNotSet)
+
+	arg.AssetType = asset.Spot
+	_, err = e.WebsocketSubmitOrder(t.Context(), arg)
+	require.ErrorIs(t, err, order.ErrSideIsInvalid)
+
+	arg.Side = order.Sell
+	_, err = e.WebsocketSubmitOrder(t.Context(), arg)
+	require.ErrorIs(t, err, order.ErrTypeIsInvalid)
+
+	arg.Type = order.TrailingStop
+	_, err = e.WebsocketSubmitOrder(t.Context(), arg)
+	require.ErrorIs(t, err, order.ErrAmountIsInvalid)
+
+	arg.Amount = 1
+	_, err = e.WebsocketSubmitOrder(t.Context(), arg)
+	require.ErrorIs(t, err, order.ErrUnsupportedOrderType)
+
+	arg.Type = order.Limit
+	_, err = e.WebsocketSubmitOrder(t.Context(), arg)
+	require.ErrorIs(t, err, order.ErrPriceMustBeSetIfLimitOrder)
+
+	arg.AssetType = asset.Futures
+	arg.Price = 10
+	_, err = e.WebsocketSubmitOrder(t.Context(), arg)
+	require.ErrorIs(t, err, asset.ErrNotSupported)
+
+	e := new(Exchange)
+	require.NoError(t, testexch.Setup(e), "Test instance Setup must not error")
+
+	if mockTests {
+		t.Skip(websocketMockTestsSkipped)
 	}
+	sharedtestvalues.SkipTestIfCredentialsUnset(t, e, canManipulateRealOrders)
+	e.setAPICredential(apiKey, apiSecret)
+	require.True(t, e.Websocket.CanUseAuthenticatedEndpoints(), "CanUseAuthenticatedEndpoints must return true")
+
+	testexch.SetupWs(t, e)
+	result, err := e.WebsocketSubmitOrder(generateContext(t), &order.Submit{
+		Exchange:    e.Name,
+		Pair:        spotTradablePair,
+		Side:        order.Buy,
+		Type:        order.Market,
+		Price:       10,
+		QuoteAmount: 1000000,
+		AssetType:   asset.Spot,
+	})
+	assert.NoError(t, err)
+	assert.NotNil(t, result)
+
+	result, err = e.WebsocketSubmitOrder(generateContext(t), &order.Submit{
+		Exchange:     e.Name,
+		Pair:         spotTradablePair,
+		Side:         order.Sell,
+		Type:         order.Limit,
+		TriggerPrice: 11,
+		Price:        10,
+		Amount:       1,
+		ClientID:     "hi",
+		AssetType:    asset.Spot,
+	})
+	assert.NoError(t, err)
+	assert.NotNil(t, result)
+}
+
+func TestWebsocketCancelOrder(t *testing.T) {
+	t.Parallel()
+	e := new(Exchange)
+	require.NoError(t, testexch.Setup(e), "Test instance Setup must not error")
```

Comment:
You will need to set the API credentials here. Please ensure all tests that shadow `e` set credentials when credentials are required